transport_analysis/viscosity.py

```diff
@@ -0,0 +1,191 @@
+        The frame to stop at for the analysis.
+    step : Optional[int]
+        Number of frames to skip between each analyzed frame.
+    n_frames : int
+        Number of frames analysed in the trajectory.
+    n_particles : int
+        Number of particles viscosity was calculated over.
+    """
+
+    def __init__(
+        self,
+        atomgroup: "AtomGroup",
+        vol_avg: float,
+        temp_avg: float = 300.0,
+        dim_type: str = "xyz",
+        **kwargs,
+    ) -> None:
+        # the below line must be kept to initialize the AnalysisBase class!
+        super().__init__(atomgroup.universe.trajectory, **kwargs)
+        # after this you will be able to access `self.results`
+        # `self.results` is a dictionary-like object
+        # that can should used to store and retrieve results
+        # See more at the MDAnalysis documentation:
+        # https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html?highlight=results#MDAnalysis.analysis.base.Results
+
+        if isinstance(atomgroup, UpdatingAtomGroup):
+            raise TypeError(
+                "UpdatingAtomGroups are not valid for viscosity computation"
+            )
+
+        # args
+        self.dim_type = dim_type
+        self._parse_dim_type()
+        # self.fft = fft # consider whether fft is possible later
+
+        # local
+        self.atomgroup = atomgroup
+        self.n_particles = len(self.atomgroup)
+
+    def _prepare(self):
+        """
+        Set up viscosity, mass, velocity, and position arrays
+        before the analysis loop begins
+        """
+        # 2D viscosity array of frames x particles
+        self.results.visc_by_particle = np.zeros(
+            (self.n_frames, self.n_particles)
+        )
+
+        self._mass_array = self.atomgroup.masses
+        self._mass_reshape = self._mass_array.reshape(
+            1, len(self._mass_array, 1)
+        )
+
+        # 3D arrays of frames x particles x dimensionality
+        # for velocities and positions
+        self._velocity_array = np.zeros(
+            (self.n_frames, self.n_particles, self.dim_fac)
+        )
+
+        self._position_array = np.zeros(
+            (self.n_frames, self.n_particles, self.dim_fac)
+        )
+        # self.results.timeseries not set here
+
+    def _parse_dim_type(self):
+        r"""Sets up the desired dimensionality of the calculation."""
+        keys = {
+            "x": [0],
+            "y": [1],
+            "z": [2],
+            "xy": [0, 1],
+            "xz": [0, 2],
+            "yz": [1, 2],
+            "xyz": [0, 1, 2],
+        }
+
+        self.dim_type = self.dim_type.lower()
+
+        try:
+            self._dim = keys[self.dim_type]
+        except KeyError:
+            raise ValueError(
+                "invalid dim_type: {} specified, please specify one of xyz, "
+                "xy, xz, yz, x, y, z".format(self.dim_type)
+            )
+
+        self.dim_fac = len(self._dim)
```

Comment:
Could we make this a static method that returns dim_type and dim_fac?

Generally, I am of the opinion that it's always best to use static methods when possible. It makes it much easier to inspect the function call and understand what's happening.

Reply:
So do you mean we would set the instance variables `self._dim` and `self.dim_fac` in `__init__()`?

Reply:
Oh, I see what you mean now. Missed the comment above, my bad!